File: modules/secure-recorder/android/src/main/java/expo/modules/securerecorder/PermissionManager.kt
package expo.modules.securerecorder

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import kotlinx.coroutines.delay

/**
 * Permission management interface
 * Matches iOS PermissionManager protocol for cross-platform consistency
 */
interface PermissionManager {
  fun has(): Boolean
  suspend fun request(): Boolean
}

/**
 * Android permission manager implementation
 * Matches IOSPermissionManager interface for cross-platform consistency
 */
class AndroidPermissionManager(private val context: Context) : PermissionManager {
  override fun has(): Boolean {
    return ContextCompat.checkSelfPermission(
      context,
      Manifest.permission.RECORD_AUDIO
    ) == PackageManager.PERMISSION_GRANTED
  }

  override suspend fun request(): Boolean {
    // Note: In Expo modules, permission requests should be handled via expo-permissions
    // For consistency with iOS async signature, we make this suspend
    // In practice, this would integrate with expo-permissions module
    delay(0) // Yield to allow cancellation
    return has()
  }
}

File: modules/secure-recorder/ios/PermissionManager.swift
import Foundation
import AVFoundation

/**
 * Permission management protocol
 */
protocol PermissionManager {
  func has() -> Bool
  func request() async -> Bool
}

/**
 * iOS/iPadOS permission manager implementation
 * Platform differences handled via build configuration
 */
class IOSPermissionManager: PermissionManager {
  func has() -> Bool {
    return AVAudioSession.sharedInstance().recordPermission == .granted
  }
  
  func request() async -> Bool {
    return await withCheckedContinuation { continuation in
      AVAudioSession.sharedInstance().requestRecordPermission { granted in
        continuation.resume(returning: granted)
      }
    }
  }
}
